api-extractor/src/definitions/ApiItem.ts

```diff
@@ -399,6 +419,150 @@ abstract class ApiItem {
+   * specified node and analyzes each one.
+   */
+  protected collectTypeReferences(node: ts.Node): void {
+    if (node.kind === ts.SyntaxKind.Block || (node.kind >= 262 && node.kind <= 291)) {
+      // Don't traverse into code blocks or JSDoc items; we only care about the function signature
+      return;
+    }
+
+    if (node.kind === ts.SyntaxKind.TypeReference) {
+      const typeReference: ts.TypeReferenceNode = node as ts.TypeReferenceNode;
+      this._processTypeReference(typeReference);
+    }
+
+    // Recurse the tree
+    for (const childNode of node.getChildren()) {
+      this.collectTypeReferences(childNode);
+    }
+  }
+
+  /**
+   * This is a helper for collectTypeReferences().  It analyzes a single TypeReferenceNode.
+   */
+  private _processTypeReference(typeReferenceNode: ts.TypeReferenceNode): void {
+    const symbol: ts.Symbol = this.extractor.typeChecker.getSymbolAtLocation(typeReferenceNode.typeName);
+    if (!symbol) {
+      // Is this bad?
+      return;
+    }
+
+    if (symbol.flags & ts.SymbolFlags.TypeParameter) {
+      // Don't analyze e.g. "T" in "Set<T>"
+      return;
+    }
+
+    const typeName: string = typeReferenceNode.typeName.getText();
+
+    let currentSymbol: ts.Symbol = symbol;
+    while (true) {
+      const importTypeReference: IImportTypeReference = this._parseImportStatement(currentSymbol);
+      if (importTypeReference) {
+        if (importTypeReference.packageName.substr(0, 2) !== './') {
+          // [CASE 1] External/Resolved
+          // This is a reference to a type from an external package, and it was resolved.
+          return;
+        }
+      }
+
+      if (!(currentSymbol.flags & ts.SymbolFlags.Alias)) {
+        break;
+      }
+      const currentAlias: ts.Symbol = this.typeChecker.getAliasedSymbol(currentSymbol);
+      if (!currentAlias || currentAlias === currentSymbol) {
+        break;
+      }
+      currentSymbol = currentAlias;
+    }
+
+    if (!currentSymbol.declarations || !currentSymbol.declarations.length) {
+      // This is a degenerate case that happens sometimes
+      return;
+    }
+
+    // We reached the actual type declaration without tracing into an external package
+    const sourceFile: ts.SourceFile = currentSymbol.declarations[0].getSourceFile();
+
+    // Is the type from a location that we don't care about?
+    if (/[\\/]node_modules[\\/]typescript[\\/]/i.test(sourceFile.fileName)
+      || /[\\/]typings[\\/]/i.test(sourceFile.fileName)) {
+      // Yes, ignore this
+      return;
+    }
+
+    if (!this.extractor.isLocalSourceFile(sourceFile)) {
+      // [CASE 2] External/Unresolved
+      // The type is apparently from an external package, however our heuristic above
+      // wasn't able to find the import statement.  This happens due to limitations of
+      // the typeChecker.getAliasedSymbol() API; we should try to improve it.
+      this.reportWarning(`Unable to resolve external type reference for "${typeName}"`);
```

Comment:
In sp-client, this warning is being raised for React references (e.g. React.ReactElement). How can we change the heuristic to catch this?